crates/wasmtime/src/compile/code_builder.rs

```diff
@@ -177,10 +206,367 @@ impl<'a> CodeBuilder<'a> {
+                        )
+                    })?;
+                }
+                BytesOrFile::Bytes(bytes) => {
+                    std::fs::write(&path, &bytes)
+                        .with_context(|| format!("failed to write to file: {}", path.display()))?;
+                }
+            }
+
+            config
+                .dependencies
+                .insert(name.to_string(), wasm_compose::config::Dependency { path });
+        }
+
+        let composer = wasm_compose::composer::ComponentComposer::new(&main_wasm_path, &config);
+        let composed = composer.compose()?;
+        Ok(composed.into())
+    }
+
+    /// Check that the main Wasm doesn't import unsafe intrinsics, keeping the
+    /// TCB to just the compile-time builtins' implementation.
+    ///
+    /// Returns the Wasm's top-level instance imports for `wasm-compose`
+    /// configuration.
+    #[cfg(feature = "compile-time-builtins")]
+    fn check_imports_for_compile_time_builtins<'c>(
+        &self,
+        main_wasm: &'c [u8],
+    ) -> Result<crate::hash_set::HashSet<&'c str>, Error> {
+        let intrinsics_import = self.unsafe_intrinsics_import.as_deref().ok_or_else(|| {
+            anyhow!("must configure the unsafe-intrinsics import when using compile-time builtins")
+        })?;
+
+        let mut instance_imports = crate::hash_set::HashSet::new();
+        let parser = wasmparser::Parser::new(0);
+        let mut level = 0;
+
+        for payload in parser.parse_all(main_wasm) {
+            match payload? {
+                wasmparser::Payload::ComponentImportSection(imports) => {
+                    if level > 0 {
+                        continue;
+                    }
+                    for imp in imports.into_iter() {
+                        let imp = imp?;
+                        // Ideally we would simply choose a new import name that
+                        // doesn't conflict with the main Wasm's imports and
+                        // plumb that through to the compile-time builtins
+                        // regardless of the import name that they use, but
+                        // unfortunately the `wasm-compose` API is not powerful
+                        // enough for us to do all that.
+                        ensure!(
+                            imp.name.0 != intrinsics_import,
+                            "main Wasm cannot import the unsafe intrinsics (`{intrinsics_import}`) \
+                             when using compile-time builtins"
+                        );
+
+                        if let wasmparser::ComponentTypeRef::Instance(_) = imp.ty {
+                            instance_imports.insert(imp.name.0);
+                        }
+                    }
+                }
+                wasmparser::Payload::ModuleSection { .. } => {
+                    level += 1;
+                }
+                wasmparser::Payload::ComponentSection { .. } => {
+                    level += 1;
+                }
```

Comment:
You could replace these by looking for `Payload::Version` too and then change the level condition to "level 1 == do the thing, everything else == skip the thing"